pow/src/main/java/tech/pegasys/artemis/pow/Eth1MinGenesisTimeBlockFinder.java

```diff
@@ -0,0 +1,280 @@
+              }
+              return firstValidBlockFuture;
+            })
+        .thenAccept(this::publishFirstValidBlock)
+        .exceptionallyCompose(
+            err -> {
+              LOG.debug(
+                  "Eth1MinGenesisTimeBlockFinder failed to find first valid block. Retry in "
+                      + Constants.ETH1_MIN_GENESIS_TIME_BLOCK_RETRY_TIMEOUT
+                      + " seconds",
+                  err);
+
+              return asyncRunner.runAfterDelay(
+                  this::findAndPublishFirstValidBlock,
+                  Constants.ETH1_MIN_GENESIS_TIME_BLOCK_RETRY_TIMEOUT,
+                  TimeUnit.SECONDS);
+            });
+  }
+
+  private void publishFirstValidBlock(EthBlock.Block block) {
+    minGenesisTimeBlockEventChannel.onMinGenesisTimeBlock(
+        new MinGenesisTimeBlockEvent(
+            UnsignedLong.valueOf(block.getTimestamp()),
+            UnsignedLong.valueOf(block.getNumber()),
+            Bytes32.fromHexString(block.getHash())));
+  }
+
+  /**
+   * Find first estimationBlock that has timestamp that is greater than MIN_GENESIS_TIME
+   *
+   * @param estimationBlock estimationBlock that will be used for estimation
+   * @return first valid block in history
+   */
+  private SafeFuture<EthBlock.Block> findFirstValidBlockInHistory(EthBlock.Block estimationBlock) {
+    UnsignedLong estimatedFirstValidBlockNumber =
+        getEstimatedFirstValidBlockNumber(estimationBlock, Constants.SECONDS_PER_ETH1_BLOCK);
+    return eth1Provider
+        .getEth1BlockFuture(estimatedFirstValidBlockNumber)
+        .thenCompose(
+            block -> {
+              SafeFuture<EthBlock.Block> firstValidBlock;
+              switch (compareBlockTimestampToMinGenesisTime(block)) {
+                case 1:
+                  // If block timestamp is greater than min genesis time
+                  // explore blocks downwards
+                  firstValidBlock = exploreBlocksDownwards(block);
+                  break;
+                case 0:
+                  firstValidBlock = SafeFuture.completedFuture(block);
+                  break;
+                case -1:
+                  // If block timestamp is less than min genesis time
+                  // explore blocks upwards
+                  firstValidBlock = exploreBlocksUpwards(block);
+                  break;
+                default:
+                  throw new IllegalStateException(
+                      "Unexpected value: "
+                          + calculateCandidateGenesisTimestamp(block.getTimestamp())
+                              .compareTo(Constants.MIN_GENESIS_TIME));
+              }
+              return firstValidBlock;
+            });
+  }
+
+  private SafeFuture<EthBlock.Block> exploreBlocksDownwards(EthBlock.Block previousBlock) {
+    UnsignedLong previousBlockNumber = UnsignedLong.valueOf(previousBlock.getNumber());
+    UnsignedLong newBlockNumber = previousBlockNumber.minus(UnsignedLong.ONE);
```

Comment:
Probably need to handle the case where we reach the ETH1 genesis block and are still after min genesis time.

Reply:
done.